api/agent/agent.go

```diff
@@ -563,143 +483,133 @@ func (s *hotSlot) exec(ctx context.Context, call *call) error {
 	case err := <-s.errC: // error from container
 		return err
 	case err := <-errApp: // from dispatch
+		if err != nil && models.IsAPIError(err) && s.fatalErr != nil {
+			// TODO: narrow this down to StatusBadGateway, we do not want to accidentally
+			// shutdown container due to client side errors. Alternatively Dispatch()
+			// should tell us if the error is from the container or not.
+			// TODO: we can skip this for Cold containers, but let's not optimize Cold path
+			s.fatalErr = err
+		}
```

Comment:
this ends up being a pretty big behavior change, didn't see a note about it so just making sure this is intended

Reply:
We briefly discussed this, we failed to parse the protocol. Before this commit, the logic here tried to recover by draining the pipes. (50 msec or 100MB read) But we said, this is too difficult and not certain to fix the pipes and adds latency. It's best to kill the container when we go into such inconsistent state. (But I might have misunderstood the comm.)

Reply:
commit, I am referring to previous commit in this PR.

Reply:
when I made the comment, it seemed to also encompass timeout / cancel and 'oversize output' but neither of those cases are handled here (the latter not being upstream, the former not a 'models.APIError') -- though we probably should handle them similarly, this is ok

Reply:
so, ok to shutdown in case of protocol errors, right? I think dispatch() probably should return two different errors, one for errors that we want to log or attempt to relay to client, etc. The other for errors related to container, which warrants a shutdown. The distinction is not that clear here. But regardless, I expect some post-cleanup work (probably outside of scope of this PR)